src/storage/disk_table.cc

```diff
@@ -329,16 +337,69 @@ void DiskTable::SchedGc() {
 void DiskTable::GcHead() {
     uint64_t start_time = ::baidu::common::timer::get_micros() / 1000;
     auto inner_indexs = table_index_.GetAllInnerIndex();
+    auto cf_hs(cf_hs_);
     for (const auto& inner_index : *inner_indexs) {
         uint32_t idx = inner_index->GetId();
         rocksdb::ReadOptions ro = rocksdb::ReadOptions();
         const rocksdb::Snapshot* snapshot = db_->GetSnapshot();
         ro.snapshot = snapshot;
         // ro.prefix_same_as_start = true;
         ro.pin_data = true;
-        rocksdb::Iterator* it = db_->NewIterator(ro, cf_hs_[idx + 1]);
-        it->SeekToFirst();
         const auto& indexs = inner_index->GetIndex();
+
+        size_t deleted_num = 0;
+        // kWaiting -> kDeleting -> kDeleted
+        for (const auto& cur_index : indexs) {
+            switch (cur_index->GetStatus()) {
+                case IndexStatus::kWaiting:
+                    cur_index->SetStatus(IndexStatus::kDeleting);
+                    break;
+                case IndexStatus::kDeleting:
+                    if (indexs.size() == 1) {
+                        rocksdb::ColumnFamilyOptions cfo;
+
+                        if (storage_mode_ == ::openmldb::common::StorageMode::kSSD) {
+                            cfo = rocksdb::ColumnFamilyOptions(ssd_option_template);
+                        } else {
+                            cfo = rocksdb::ColumnFamilyOptions(hdd_option_template);
+                        }
+
+                        cfo.comparator = &cmp_;
+                        cfo.prefix_extractor.reset(new KeyTsPrefixTransform());
+
+                        if (cur_index->GetTTLType() == ::openmldb::storage::TTLType::kAbsoluteTime ||
+                            cur_index->GetTTLType() == ::openmldb::storage::TTLType::kAbsOrLat) {
+                            cfo.compaction_filter_factory = std::make_shared<AbsoluteTTLFilterFactory>(inner_index);
+                        }
+
+                        rocksdb::ColumnFamilyHandle* handle;
+                        db_->DropColumnFamily(cf_hs->at(idx + 1));
+                        db_->DestroyColumnFamilyHandle(cf_hs->at(idx + 1));
+                        db_->CreateColumnFamily(cfo, cur_index->GetName(), &handle);
+                        // modify element of cf_hs_
+                        {
+                            std::lock_guard<std::mutex> guard(cf_hs_mutex_);
+                            cf_hs_->at(idx + 1) = handle;
+                        }
+                        deleted_num++;
+                    } else {
+                        // do nothing following mem_table
```

Comment:
when are the data of these indexes deleted after deleting of all the indexes?

Reply:
See here https://github.com/4paradigm/OpenMLDB/blob/35a418eccdaa6021bf2e4b808ce737f0b0e1984a/src/storage/disk_table.cc#L375-L377
After dropping the column family, the data inside it will be deleted by `rockdb` at some point. Say [here](http://mysql.taobao.org/monthly/2018/06/09/) the "some point" is when the reference count to the column family decreases to 0.

https://github.com/facebook/rocksdb/issues/1295
https://github.com/facebook/rocksdb/issues/5006

Reply:
No. Here I mean when the condition `if (indexs.size() == 1)` will be true. I didn't find any code you remove or change the size of inner index.

Reply:
If I get this right, `MemTable` do not delete the `inner_index` either and reserve it for future reuse.

https://github.com/4paradigm/OpenMLDB/blob/35a418eccdaa6021bf2e4b808ce737f0b0e1984a/src/storage/disk_table.cc#L1177-L1184

Reply:
If I get this right, `MemTable` do not delete elements of the `inner_index` either and keep it for future reuse.

https://github.com/4paradigm/OpenMLDB/blob/35a418eccdaa6021bf2e4b808ce737f0b0e1984a/src/storage/disk_table.cc#L1177-L1184

Reply:
Yes. `MemTable` has a bug for this case #2680

There is a PR to fix the bug: #2723, but still not complete. 

For `DiskTable`, it is not easy to only delete the data in one single index if the size of its parent inner index > 1. Maybe you can put an issue to take note of this bug. We can fix it in another PR.

Reply:
Got it.

Reply:
#2748